### ISSUE TITLE
How to get NULL with ZeroOrOne and Decimal parser

### ISSUE THREAD

**Issue by loudenvier** [sebastienros/parlot#81]
I'm trying to parse a decimal which is optional and with a cardinality of zero or one. I've tried the following:

```csharp
var dec = ZeroOrOne(Terms.Decimal());
var parsed = dec.Parse("");
Console.WriteLine($"parsed: {parsed}");
```
Which outputs: `parsed: 0`

I can't differentiate it from an actual "0" literal! I would expect that `ZeroOrOne` resulted in null in case of no match.

Is there a way to differentiate "" from "0" when parsing this? Currently `dec.Parse("") == dec.Parse("0")`.

**Comment by hishamco** [comment]
IMHO `""` and `"0"` should be the same in terms of decimal, since there are no nullable types supported

**Comment by sebastienros** [comment]
ZeroOrOne will always work and return the default value of the parsed statement. `0` being the default. And you can't return a `decimal?` since `Terms.Decimal()` returns `decimal` (unfortunately for you).

The solution I found is to read a decimal, and otherwise return `null` like this:

```c#
var dec = Terms.Decimal().Then<decimal?>(x => x).Or(Empty<decimal?>(null));
```

**Comment by sebastienros** [comment]
`Empty` is a parser that always returns a value, so if a decimal is not found, it will evaluate the `Or` and then return `null`. The `Then` is here to cast the expression to allow returning a nullable decimal.

**Comment by loudenvier** [comment]
> IMHO `""` and `"0"` should be the same in terms of decimal, since there are no nullable types supported

I don't think this is the best design decision. Parsing `""` as a decimal should fail (and I believe it indeed fails). I'm getting 0 here because `ZeroOfOne` returns the default for the parser "inside". Ideally a `decimal?` parser could exist, or the `decimal` parser could be changed to return `null` (but this will certainly break code). 

It's hard to parse when you can't differentiate "" from "0". I didn't write my complete usage here because I was trying to be succinct. But for completeness what I'm trying to do is to parse an RGB with optional Alpha component as bellow:

```csharp
var openParen = Terms.Char('(');
var closeParen = Terms.Char(')');
var colon = Terms.Char(',');
var colorArgs = Terms.Integer().AndSkip(colon)
    .And(Terms.Integer()).AndSkip(colon)
    .And(Terms.Integer())
    .And(ZeroOrOne(colon.SkipAnd(Terms.Decimal())));
var colorArguments = Between(openParen, colorArgs, closeParen);
var color = Terms.Text("rgb", true).SkipAnd(colorArguments)
    .Then(c => Color.FromArgb((int)(c.Item4 * 255), (int)c.Item1, (int)c.Item2, (int)c.Item3));

var compiled = color.Compile();
var color1 = compiled.Parse("rgb(100, 200, 255, 0.5)");
var color2 = compiled.Parse("rgb(100, 200, 255)");

Console.WriteLine($"color1: {color1}\r\ncolor2: {color2}");
```
Which results in: 
```
color1: Color [A=127, R=100, G=200, B=255]
color2: Color [A=0, R=100, G=200, B=255]
```

Unfortunately since ZeroOrOne together with Term.Decimal and "" returns 0, I can't identify when the last parameter was missing and use an alpha of 1 instead of 0. If it resulted in "null" that would be trivial.

Using Sébastien's suggestion it work, but compilation fails:
```csharp
var colorArgs = Terms.Integer().AndSkip(colon)
    .And(Terms.Integer()).AndSkip(colon)
    .And(Terms.Integer())
    .And(ZeroOrOne(colon.SkipAnd(Terms.Decimal().Then<decimal?>(x => x).Or(Empty<decimal?>(null)))));
...
//var compiled = color.Compile();
var color1 = color.Parse("rgb(100, 200, 255, 0.5)");
var color2 = color.Parse("rgb(100, 200, 255)");
```
If you try to compile it you get this error: _System.ArgumentException: 'Expression of type 'System.Object' cannot be used for assignment to type 'System.Nullable`1[System.Decimal]''_

If you don't bother compiling, it work as expected and returns null for the missing "decimal" and I can then detect it when building the "Color". The modified code yields this result:
```
color1: Color [A=127, R=100, G=200, B=255]
color2: Color [A=1, R=100, G=200, B=255]
```

Which is the expected result and works for me! But we loose a few nanoseconds per parsing, and it's probably a bug in the compilation process... but I can live with that!
